### ISSUE TITLE
[resolver] Score missing responses by elapsed time

### ISSUE THREAD

**Issue by patrick-ogrady** [commonwarexyz/monorepo#512]
## Background

The original concern was that a peer could answer a request without useful data and avoid any performance penalty. That behavior is fixed: the resolver treats an error/missing response as a failed request, retries it elsewhere, and does not penalize a response when the consumer has already marked the item as ignored.

Today, however, every missing response receives the same full-timeout sample, even if the peer promptly reports that it does not have the data.

## Remaining improvement

Score a missing/error response by the elapsed time between request and response, while capping the resulting score at the configured timeout.

This should:

- Penalize repeated unhelpful responses, so they cannot remain preferred.
- Prefer a prompt negative response to a timeout, so peers are not incentivized to stay silent when they lack the item.
- Preserve the existing behavior for invalid data: block the sender and retry.
- Preserve the existing behavior for `Outcome::Ignored`: do not score a response once the consumer no longer needs the item.

The update should compose with the requester's current exponential moving average. In particular, an incomplete response should contribute its observed elapsed duration, bounded so its score cannot exceed the timeout.

## Acceptance criteria

- A missing/error reply is retried and reduces the peer's selection priority.
- A prompt missing/error reply is scored less severely than a timeout.
- Repeated prompt missing/error replies still make a peer less preferred than one that supplies useful data at comparable latency.
- Tests cover missing/error, timeout, useful, invalid, and ignored response paths.

**Comment by BrendanChou** [comment]
An interesting way to try and score this is to:
- take the time that the peer took to respond
- "append" this time on to the previous request

That is, try and basically find the amount of network time that we are waiting for a peer between legitimate, contentful responses. The following peers would score the same:
- Peer A responds with a valid response after 400ms every time
- Peer B responds always in 200ms, but half the time gives no answer and half the time gives the valid response

This penalizes peers that continually give empty response, but does "incentivize" peers to respond quickly if they don't have the answer (rather than dropping it and letting it time out).

There are several ways to implement this, some more "correct" than others. A way to do it without introducing any additional state would be to simply add another method to the `Requester` that is similar to `resolve()` and `timeout()`. Let's call the function `incomplete()` for now. `incomplete()` should increase the peer's score by `learning_rate * response_duration`. However peers should not be "punished" more than they would be a for a timeout, so the maximum increase should be `(timeout_duration - current_score) * learning_rate`

The learning rate is currently hard-coded to `0.5`, so the full formula for the score increase would be `min(response_duration, timeout_duration - current_score) / 2`. This can be achieved by:


```
pub fn incomplete(&mut self, request: Request<C::PublicKey>) {
    let past = self.participants.get(request.participant).unwrap_or(self.initial);
    let elapsed = self
        .context
        .current()
        .duration_since(request.start)
        .unwrap_or_default();
    let next = past + elapsed;
    self.update(request.participant, min(next, self.timeout));
}
```

**Comment by malik672** [comment]
a proof of history is technically the best option here tbh
or a time limit

**Comment by BrendanChou** [comment]
> a proof of history is technically the best option here tbh or a time limit

Can you expand on this a bit?

**Comment by malik672** [comment]
> > a proof of history is technically the best option here tbh or a time limit
> 
> Can you expand on this a bit?

  have an average worst time upper bound we can simply penalize both empty response and malicious peers

**Comment by Sertug17** [comment]
After reviewing the current code, `pop_missing()` in `resolver/src/p2p/fetcher.rs` already applies a timeout penalty via update_performance(&req.peer, self.timeout)` when a peer returns an empty/error response. This appears to address the concern raised here. Should this issue be closed?

**Comment by BrendanChou** [comment]
Rescoped this to track the remaining peer-selection refinement.

The original bypass is fixed: missing/error replies are no longer free, and ignored responses are not scored. The remaining problem is that an immediate negative reply receives the same full-timeout penalty as silence. This issue will stay open to score missing replies by their observed elapsed time instead.

**Comment by Sertug17** [comment]
I'd like to take a look at this. Planning to implement the incomplete() method on Requester that scores missing/error responses by elapsed time (capped at timeout), as outlined in the comments above. @BrendanChou
